### PR TITLE
Convert JVM target to use KMP

### PR DESCRIPTION
**Pull request by krzema12**

### CODE REVIEW

File: src/jvmMain/kotlin/com/charleskorn/kaml/Yaml.kt [hunk 1]
```diff
@@ -1,119 +0,0 @@
...
-
-public actual class Yaml actual constructor(
-    override val serializersModule: SerializersModule,
-    public actual val configuration: YamlConfiguration,
-) : StringFormat {
-    override fun <T> decodeFromString(deserializer: DeserializationStrategy<T>, string: String): T {
-        return decodeFromReader(deserializer, StringReader(string))
-    }
-
-    public fun <T> decodeFromStream(deserializer: DeserializationStrategy<T>, source: InputStream, charset: Charset = Charsets.UTF_8): T {
```
Comment (krzema12) [inline src/jvmMain/kotlin/com/charleskorn/kaml/Yaml.kt:45]: I removed too much from the JVM source set. If we want to preserve the JVM API, we need to revive some of this code.

Comment (charleskorn) [reply src/jvmMain/kotlin/com/charleskorn/kaml/Yaml.kt:45]: I'd like to preserve the JVM API, so yes, please keep whatever's required to maintain the API as-is.

Comment (krzema12) [reply src/jvmMain/kotlin/com/charleskorn/kaml/Yaml.kt:45]: Sure, working on it!

**Comment by krzema12** [comment]
I have to drop working on this, sorry! Feel free to take over.

**Comment by westnordost** [comment]
Could you summarize what is still missing in this PR? Do you think it would require adjustments in snakeyaml-engine-kmp itself?

**Comment by krzema12** [comment]
@westnordost for sure the JVM-specific, not sure what more. I recall not being able to get past some failures in unit tests. If some changes are needed in snakeyaml-engine-kmp, I'm the owner of it, so I'll be glad to help.

**Comment by aSemy** [comment]
I've picked this up and I've managed to get the existing tests passing 🎉.

@krzema12 Would you be able to merge these PRs?

* https://github.com/krzema12/kaml/pull/2
* https://github.com/krzema12/snakeyaml-engine-kmp/pull/135

**Comment by krzema12** [comment]
Awesome, thanks @aSemy! Next step, on me: release snakeyaml-engine-kmp with your fix so that we can use it in this PR.

**Comment by krzema12** [comment]
https://github.com/krzema12/snakeyaml-engine-kmp/releases/tag/v2.7.2

File: src/jvmMain/kotlin/com/charleskorn/kaml/JvmYamlReading.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    deserializer: DeserializationStrategy<T>,
+    source: InputStream,
+    charset: Charset = Charsets.UTF_8,
+): T {
+    return decodeFromSource(
+        deserializer,
+        source.source(),
+        charset,
+    )
+}
```
[the remark this answers is not in the document]
Comment (westnordost) [reply src/jvmMain/kotlin/com/charleskorn/kaml/JvmYamlReading.kt]: Just my five cents, as a user of kaml: Unlike - as far as I know - kotlinx-serialization-json, kaml's API offers to decode from and to streams (java InputStream / okio Source). I think it is no surprise that kotlinx-serialization-json does not offer such an API, for currently - as far as I know -, there is no official Kotlin multiplatform library that offers a file access or even just a input/output stream API.

Without reading/writing to an input/output byte stream but only handling Strings, the official json serialization library does not have to deal with charsets at all.

While not even the official kotlinx serialization libraries offer this feature, it seems unnecessary to go out of one's way to support an API that is itself not supported by the libraries one builds upon (in this case, for input/output byte stream reading, okio).

There is the official [`kotlinx-io-core`](https://github.com/Kotlin/kotlinx-io) kmp library that should deal with byte streams (et al), but it is still in its experimental phase (and at least currently, depends on okio anyway).

In my opinion, as long as no official stable kmp library exists that offers reading and writing byte array streams, it would be best to go even one step further and actually only support serialization from and to Strings. (Now, whether or not okio could be considered official / de-facto standard, hm...)

[the remark this answers is not in the document]
Comment (aSemy) [reply src/jvmMain/kotlin/com/charleskorn/kaml/JvmYamlReading.kt]: Kotlinx Serialization does support Java and Okio streams :)

* https://github.com/Kotlin/kotlinx.serialization/blob/v1.6.3/formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt
* https://github.com/Kotlin/kotlinx.serialization/blob/v1.6.3/formats/json-okio/commonMain/src/kotlinx/serialization/json/okio/OkioStreams.kt

It doesn't support different charsets though.

[the remark this answers is not in the document]
Comment (westnordost) [reply src/jvmMain/kotlin/com/charleskorn/kaml/JvmYamlReading.kt]: Oh!

**Comment by krzema12** [comment]
@charleskorn could you review please?

File: src/commonMain/kotlin/com/charleskorn/kaml/Yaml.kt [hunk 3]
```diff
@@ -18,20 +18,109 @@
...
+        return node
+    }
+
+    public fun <T> encodeToSink(
+        serializer: SerializationStrategy<T>,
+        value: T,
+        sink: Sink,
+    ) {
+        encodeToBufferDataWriter(serializer, value).use { writer ->
+            writer.readAll(sink)
```
Comment (charleskorn) [inline src/commonMain/kotlin/com/charleskorn/kaml/Yaml.kt:85]: In the case of writing to a `Sink`, couldn't we write directly to the sink rather than through a buffer? We'd need another implementation of `StreamDataWriter` that writes directly to a `Sink`

Comment (charleskorn) [reply src/commonMain/kotlin/com/charleskorn/kaml/Yaml.kt:85]: Alternatively, `BufferedDataWriter` could take a `Sink` (and be renamed to `SinkDataWriter`) and then we can pass the `Sink` in this method and a `Buffer` in `encodeToString`

**Review by charleskorn — COMMENTED** [review]
Sorry for the slow response to this, I've been a bit busy!

Thanks for working on this, this looks good overall.

File: build.gradle.kts [hunk 4]
```diff
@@ -80,28 +78,15 @@ kotlin {
...
-            dependencies {
-                implementation("it.krzeminski:snakeyaml-engine-kmp:2.7.3")
-                implementation("com.squareup.okio:okio:3.9.0")
-            }
-        }
     }
 }
 
-tasks.withType<KotlinCompile> {
+tasks.withType<KotlinCompile>().configureEach {
```
[the remark this answers is not in the document]
Comment (aSemy) [reply build.gradle.kts:89]: It's not strictly required, I was just tidying while I was working in this file. Using `configureEach {}` helps improve Gradle configuration speed by avoiding work.

`tasks.withType<KotlinCompile> { }` will create and configure all KotlinCompile tasks, whether they're needed or not. By adding `configureEach {}`, Gradle will avoid creating and configuring the tasks unless necessary.

https://docs.gradle.org/8.6/userguide/task_configuration_avoidance.html#sec:old_vs_new_configuration_api_overview

This isn't related to the current ticket, so if you'd like you can revert it.

File: src/jvmMain/kotlin/com/charleskorn/kaml/JvmYamlReading.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ */
+// The character encoding is not configurable, because we use Okio which doesn't support converting
+// between UTF-8 and other encodings.
+public fun <T> Yaml.decodeFromStream(
+    deserializer: DeserializationStrategy<T>,
+    source: InputStream,
+): T = decodeFromSource(
+    deserializer = deserializer,
+    source = source.source(),
+)
```
[the remark this answers is not in the document]
Comment (aSemy) [reply src/jvmMain/kotlin/com/charleskorn/kaml/JvmYamlReading.kt:39]: * Personal preference,
* adding explicit names helped when I was learning about the code (especially when there are overloaded functions),
* in an IDE I like that I can cmd+click on a parameter name to go straight to it,
* and when I was refactoring so that when I modified the args the compilation errors were more explicit.

I'd be happy to remove them, if you prefer!

File: src/commonMain/kotlin/com/charleskorn/kaml/Yaml.kt (hunk 3 above)
Comment (aSemy) [reply src/commonMain/kotlin/com/charleskorn/kaml/Yaml.kt]: Good idea, done!

I had to use a BufferedSink instead, because there's no `Sink.writeUtf8()`.

File: src/commonMain/kotlin/com/charleskorn/kaml/Yaml.kt [hunk 6]
```diff
@@ -18,20 +18,111 @@
...
+        sink.writeUtf8(str)
+    }
+
+    override fun write(str: String, off: Int, len: Int) {
+        sink.writeUtf8(string = str, beginIndex = off, endIndex = off + len)
+    }
+
+    override fun close() {
+        flush()
     }
```
Comment (aSemy) [inline src/commonMain/kotlin/com/charleskorn/kaml/Yaml.kt:127]: I'm not 100% certain flushing is necessary or important, but I'm concerned about missing data in resources-intensive situations. It probably doesn't hurt.

File: build.gradle.kts
Comment (aSemy) [inline build.gradle.kts:1]: Regarding the disabled JS tests:

```kotlin
        browser {
            testTask {
                // TODO: enable once the tests work with Kotlin/JS.
                enabled = false
            }
        }
        nodejs {
            testTask {
                // TODO: enable once the tests work with Kotlin/JS.
                enabled = false
            }
        }
```

I've implemented a workaround in Kotest, so nested tests should work.

https://github.com/kotest/kotest/pull/3913

So, please +1 and review that PR :)

File: build.gradle.kts (hunk 4 above)
[the remark this answers is not in the document]
Comment (charleskorn) [reply build.gradle.kts:89]: Makes sense, happy to keep it.

File: src/jvmMain/kotlin/com/charleskorn/kaml/JvmYamlReading.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (charleskorn) [reply src/jvmMain/kotlin/com/charleskorn/kaml/JvmYamlReading.kt:39]: Happy to keep them.

File: src/jvmMain/kotlin/com/charleskorn/kaml/JvmYamlReading.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (charleskorn) [reply src/jvmMain/kotlin/com/charleskorn/kaml/JvmYamlReading.kt]: I think it's OK to drop support for non-UTF-8 charsets for now - if anyone needs anything else, they can convert it themselves or submit a PR to add support for it 🙂

**Comment by charleskorn** [comment]
Thanks for the PR @krzema12 and @aSemy!

**Comment by westnordost** [comment]
Very cool, thanks for your work on this, @krzema12  and @aSemy !
